### ISSUE TITLE
Optimisation prioritisation problems for reward and KL in PPO training

### ISSUE THREAD

**Issue by Loreley99** [huggingface/trl#1178]
Thank you very much for developing such a cool library! But I'm having some optimisation issues with training. I am currently using mistral as a generative model for PPO. My PPO configs are set as follows:
adap_kl_ctrl: true
backward_batch_size: 64
batch_size: 512
cliprange: 0.2
cliprange_value: 0.2
compare_steps: 1
early_stopping: false
gamma: 1
global_backward_batch_size: 64
global_batch_size: 512
gradient_accumulation_steps: 4
horizon: 10,000
init_kl_coef: 0.2
is_encoder_decoder: false
is_peft_model: true
kl_penalty: "kl"
lam: 0.95
learning_rate: 0.0000141
mini_batch_size: 16
model_name: "mistralai/Mistral-7B-v0.1"
ppo_epochs: 4
ratio_threshold: 10
steps: 50,000
target: 6
target_kl: 1
total_ppo_epochs: 98
vf_coef: 0.1
world_size: 1

Currently the kl curve trend looks all right.
But I want to get a higher reward. I noticed that after step12 it looks like the model cares more about lowering the kl and stops boosting the reward. i tried to increase the batch size and adjust the learning rate, I also tried change init_kl_coef to 0.1,  but that didn't help much. Can you please provide some suggestions on how to optimise both reward and kl?
![image](https://github.com/huggingface/trl/assets/84299634/a4e19ee2-2327-4665-8d83-2101a8b6058f)

Thanks very much!

**Comment by lvwerra** [comment]
What kind of reward are you using? Maybe it is a hard reward to learn and requires slightly longer to get there. Have you tried e.g. with the imdb classifier that we use in the examples? You should see success quickly. Otherwise I'd suggest running longer and see if improves over time.

**Comment by Loreley99** [comment]
Thanks for your quick reply. The reward I'm using it's an output of multiple LM models. 
I understand that this may be more difficult to learn, but there is one point in particular that I don't understand: The model learns the reward function in the first few steps. But then it switches to optimizing KL in the last steps even though kl_coef is lower. Could you please provide some advice on this?

> What kind of reward are you using? Maybe it is a hard reward to learn and requires slightly longer to get there. Have you tried e.g. with the imdb classifier that we use in the examples? You should see success quickly. Otherwise I'd suggest running longer and see if improves over time.

**Comment by lvwerra** [comment]
You could try to reduce the KL coefficient further or train longer. I have observed in the past that the models learns a bit but goes down again and then learns in the long run. I would see if it improves after O(100) steps or more.
